Framework/src/CheckRunner.cxx

```diff
@@ -0,0 +1,325 @@
+    if (dataRef.header != nullptr && dataRef.payload != nullptr) {
+      auto moArray = ctx.inputs().get<TObjArray*>(input.binding.c_str());
+      mLogger << "Device " << mDeviceName
+              << " received " << moArray->GetEntries()
+              << " MonitorObjects from " << input.binding
+              << AliceO2::InfoLogger::InfoLogger::endm;
+
+      for (const auto& to : *moArray) {
+        std::shared_ptr<MonitorObject> mo{ dynamic_cast<MonitorObject*>(to) };
+
+        if (mo) {
+          update(mo);
+          mTotalNumberHistosReceived++;
+        } else {
+          mLogger << "The mo is null" << AliceO2::InfoLogger::InfoLogger::endm;
+        }
+      }
+    }
+  }
+
+  // Check if compliant with policy
+  auto triggeredChecks = check(mMonitorObjects);
+  store(triggeredChecks);
+  send(triggeredChecks, ctx.outputs());
+
+  // Update global revision number
+  updateRevision();
+
+  // monitoring
+  endLastObject = system_clock::now();
+  if (timer.isTimeout()) {
+    timer.reset(1000000); // 10 s.
+    mCollector->send({ mTotalNumberHistosReceived, "objects" }, o2::monitoring::DerivedMetricMode::RATE);
+  }
+}
+
+void CheckRunner::update(std::shared_ptr<MonitorObject> mo)
+{
+  mMonitorObjects[mo->getFullName()] = mo;
+  mMonitorObjectRevision[mo->getFullName()] = mGlobalRevision;
+}
+
+std::vector<Check*> CheckRunner::check(std::map<std::string, std::shared_ptr<MonitorObject>> moMap)
+{
+  mLogger << "Running " << mChecks.size() << " checks for " << moMap.size() << " monitor objects"
+          << AliceO2::InfoLogger::InfoLogger::endm;
+
+  std::vector<Check*> triggeredChecks;
+  for (auto& check : mChecks) {
+    if (check.isReady(mMonitorObjectRevision)) {
+      auto qualityObj = check.check(moMap);
+      // Check if shared_ptr != nullptr
+      if (qualityObj) {
+        triggeredChecks.push_back(&check);
+      }
+
+      // Was checked, update latest revision
+      check.updateRevision(mGlobalRevision);
+    } else {
+      mLogger << "Monitor Objects for the check '" << check.getName() << "' are not ready, ignoring" << AliceO2::InfoLogger::InfoLogger::endm;
+    }
+  }
+  return triggeredChecks;
+}
+
+void CheckRunner::store(std::vector<Check*>& checks)
+{
+  mLogger << "Storing " << checks.size() << " quality objects" << AliceO2::InfoLogger::InfoLogger::endm;
+  try {
+    for (auto check : checks) {
+      mDatabase->storeQO(check->getQualityObject());
+    }
+  } catch (boost::exception& e) {
+    mLogger << "Unable to " << diagnostic_information(e) << AliceO2::InfoLogger::InfoLogger::endm;
+  }
+
+  mLogger << "Storing updated monitor objects" << AliceO2::InfoLogger::InfoLogger::endm;
+  try {
+    // Ensure to write only once updated MOs
+    std::set<std::string> moNames;
+    for (auto check : checks) {
+      if (check->isBeautified()) {
+        // If beautified then only one MO
+        auto name = *(check->getMonitorObjectNames().begin());
+        moNames.insert(name);
+      }
+    }
```

Comment:
When are objects without checks written to database ?

Reply:
If I change the name of the object to check in basic.json at line 53, then the object is not stored in the database. This is a bug. All objects must in the end be in the database, whether they are checked or not.

Reply:
Yes. I wrongly assumed, that the MO is stored in task and here it has to be only updated.
Although shouldn't it be the case that storing of the MO is done by the producer/task? If a MO is used by multiple CheckRunners, there will be multiple copies of the same or at least multiple access to the database. The CheckRunner in this case will only override once after beautification.

Reply:
Originally, only one Checker (CheckRunner) would process a given MO and it could beautify it. As a consequence, the Checker was the last guy in the chain who could modify the MOs and it made sense to store it there. 

Tasks can, and will, be ran in parallel on many hosts thus it is not a good place to store it. 

Mergers do not get all objects as they are bypassed by the objects who do not require them. We could change this scheme but given that they are already a contention point, I would not like that. 

We could add a device dedicated to the storing of MOs, probably after the CheckRunners. They could check whether an object has already been stored (if it is exactly the same, by storing a hash or something). This means transporting more data and some extra development. Doable though, and somehow a cleaner separation of responsibilities. 

If we want to merge this PR shortly we should find a temporary solution at the very least. I will talk with Piotr this afternoon to see what is the best solution. Rafal, do you think there is a way to determine which Checker should store a given MO to avoid conflicts with the others ? If there is no beautification, the worst case is that everyone is writing the same object again and again. Bad but not catastrophic. In case of a beautification, other CheckRunners must not store at all, otherwise they would overwrite the beautified object.

Reply:
I can think of three slightly different solutions. 
1. It is ok to store all the MOs if it is a single input. If a given task does not have a checker with a single input, there could be created a StoreDevice, that would take this single input and store it.  
2. A function with the knowledge of all checks and inputs is running on every CheckRunner device. If the output is the same as the input/name in the currently processed CheckRunner, the MO is stored here. The function could be based on the same mechanism as grouping of the checks and after sorting all, the first would store. If to store a set of MOs is determined inside the CheckRunner.
3. Similar to (2), determine who stores all MOs from an input during the insfrastructureGeneration and pass an input set to the device.

All methods might have problem if the configuration changes during the run (e.g. a check is suddenly disabled and the checkrunner closes).

As for the temporary solution, I think the fastest to achieve is (3).

At which moment should the MOs be stored? Whenever they are received by the CheckRunner or whenever a set of MOs will trigger a check? A MO could potentially not be used in case of OnAll policy, if a MO will be updated faster then the other.

Reply:
Hi Rafal, 
We discussed with Piotr and we think that what you described in option 3. is the way to go, at least for now. To say it in other words, during the infrastructure generation you have all the knowledge needed to determine which checker should store which MO (either there is a single one and it should be this one or there are several and you should elect one of them). You can therefore pass this information to the corresponding checkrunners. 

We will have to see how to handle the change of configuration indeed. 

The objects should be stored after the beautification if any, otherwise it does not matter as long as it happens in a decent time. 

@knopers8 please correct me if I am wrong

Rafal, what do you think ?
Cheers,